pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,223 @@
+			// send something to ledmanager ???
+			return fmt.Errorf("getLocalProfileConfig: resp header error: %s", err)
+		}
+		getconfigCtx.localProfileReceived = true
+		return validateAndSetLocalProfile(contents, getconfigCtx)
+	}
+	return fmt.Errorf("getLocalProfileConfig: wrong response status code: %d", resp.StatusCode)
+}
+
+// filterAndPublishAppInstancesWithCurrentProfile check all app instances with currentProfile and set activate state
+func filterAndPublishAppInstancesWithCurrentProfile(getconfigCtx *getconfigContext) {
+	pub := getconfigCtx.pubAppInstanceConfig
+	items := pub.GetAll()
+	for _, c := range items {
+		config := c.(types.AppInstanceConfig)
+		oldActivate := config.Activate
+		filterAppInstanceWithCurrentProfile(&config, getconfigCtx)
+		if oldActivate != config.Activate {
+			log.Functionf("filterAppInstancesWithLocalProfile: change activate state for %s from %t to %t",
+				config.Key(), oldActivate, config.Activate)
+			if err := pub.Publish(config.Key(), config); err != nil {
+				log.Errorf("failed to publish: %s", err)
+			}
+		}
+	}
+}
+
+func filterAppInstanceWithCurrentProfile(config *types.AppInstanceConfig, getconfigCtx *getconfigContext) {
+	if getconfigCtx.currentProfile == "" {
+		log.Functionf("filterAppInstanceWithCurrentProfile(%s): empty current", config.Key())
+		// if currentProfile is empty set activate state from controller
+		config.Activate = config.ControllerActivateState
+		return
+	}
+	if len(config.ProfileList) == 0 {
+		log.Functionf("filterAppInstanceWithCurrentProfile(%s): empty ProfileList", config.Key())
+		//we have no profile in list so we should use activate state from the controller
+		config.Activate = config.ControllerActivateState
+		return
+	}
+	config.Activate = false
+	for _, p := range config.ProfileList {
+		if p == getconfigCtx.currentProfile {
+			log.Functionf("filterAppInstanceWithCurrentProfile(%s): profile form list (%s) match current (%s)",
+				config.Key(), p, getconfigCtx.currentProfile)
+			// activate app if currentProfile is inside ProfileList
+			config.Activate = true
+			return
+		}
+	}
+	log.Functionf("filterAppInstanceWithCurrentProfile(%s): no match with current (%s)",
+		config.Key(), getconfigCtx.currentProfile)
+}
+
+func writeReceivedProtoMessageLocalProfile(contents []byte) {
+	writeProtoMessage(savedLocalProfileFile, contents)
+}
+
+func parseProfile(ctx *getconfigContext, config *zconfig.EdgeDevConfig) {
+	log.Functionf("parseProfile start: globalProfile: %s currentProfile: %s", ctx.globalProfile, ctx.currentProfile)
+	ctx.globalProfile = config.GlobalProfile
+	ctx.localProfileServer = config.LocalProfileServer
+	ctx.profileServerToken = config.ProfileServerToken
+	if config.LocalProfileServer == "" {
+		// if we do not use LocalProfileServer set profile to GlobalProfile
+		ctx.currentProfile = config.GlobalProfile
+	} else {
+		// try to read saved local profile actual app may be loaded later
+		if err := readSavedLocalProfile(ctx); err != nil {
```

Comment:
Do you really want to do this on each call to parseProfile?

Reply:
Added check to not retry it on success.